### ISSUE TITLE
AddressSanitizer: memcpy-param-overlap (in openresty-1.19.9)

### ISSUE THREAD

**Issue by a-shvedov** [openresty/lua-nginx-module#2033]
Unfortunately, the error repeats again as in openresty-1.15.8. Closed [#issues1938](https://github.com/openresty/lua-nginx-module/issues/1938)
### Version
nginx version: openresty/1.19.9.1

### Platform
3.10.0-1160.59.1.el7.x86_64 #astra26+ci17 x86_64 GNU/Linux (Red Hat Enterprise Linux Server release 7.9 (Maipo))

### Run: 
```bash
ASAN_OPTIONS=log_path=/home/user/openresty-1.19.9.1/asanlog/asanlog:log_exe_name=1:verbosity=1:exitcode=0:abort_on_error=1:handle_segv=0:detect_leaks=0:symbolize=0 /usr/local/openresty/bin/openresty  -c /usr/local/openresty/nginx/conf/id_000000.conf -t
```

Config file: [id_000000.conf](https://github.com/a-shvedov/res/blob/master/fuzzing/openresty-1.19.9/id_000000.conf)
 
### Error:
```bash
=================================================================
==32924== ERROR: AddressSanitizer: memcpy-param-overlap: memory ranges [0x60620000d900,0x60620000e65b) and [0x60620000dba5, 0x60620000e900) overlap
    #0 0x7f66fa37da74 (/usr/lib64/libasan.so.0+0xea74)
    #1 0x993654 (/usr/local/openresty/nginx/sbin/nginx+0x993654)
    #2 0x997f01 (/usr/local/openresty/nginx/sbin/nginx+0x997f01)
    #3 0x4d9948 (/usr/local/openresty/nginx/sbin/nginx+0x4d9948)
    #4 0x5cd077 (/usr/local/openresty/nginx/sbin/nginx+0x5cd077)
    #5 0x4d9948 (/usr/local/openresty/nginx/sbin/nginx+0x4d9948)
    #6 0x5b1d1a (/usr/local/openresty/nginx/sbin/nginx+0x5b1d1a)
    #7 0x4d9948 (/usr/local/openresty/nginx/sbin/nginx+0x4d9948)
    #8 0x4c7be5 (/usr/local/openresty/nginx/sbin/nginx+0x4c7be5)
    #9 0x42d2cd (/usr/local/openresty/nginx/sbin/nginx+0x42d2cd)
    #10 0x7f66f8627554 (/usr/lib64/libc-2.17.so+0x22554)
    #11 0x435cdc (/usr/local/openresty/nginx/sbin/nginx+0x435cdc)
0x60620000d900 is located 0 bytes inside of 4096-byte region [0x60620000d900,0x60620000e900)
allocated by thread T0 here:
    #0 0x7f66fa385119 (/usr/lib64/libasan.so.0+0x16119)
    #1 0x54ad26 (/usr/local/openresty/nginx/sbin/nginx+0x54ad26)
0x60620000dba5 is located 677 bytes inside of 4096-byte region [0x60620000d900,0x60620000e900)
allocated by thread T0 here:
    #0 0x7f66fa385119 (/usr/lib64/libasan.so.0+0x16119)
    #1 0x54ad26 (/usr/local/openresty/nginx/sbin/nginx+0x54ad26)
==32924== ABORTING
 ```

**Comment by zhuizhuhaomeng** [comment]
do you build openresty by yourself or install from official repo?

**Comment by zhuizhuhaomeng** [comment]
this bug has been fixed  in commit  3f1c6834ee018c057de3d96ca20c32fd87abe9c9

**Comment by a-shvedov** [comment]
> do you build openresty by yourself or install it from the official repo?

I use the source-tarball from the official site openresty.
```
([openresty-1.19.9.1.tar.gz](https://openresty.org/download/openresty-1.19.9.1.tar.gz)   
5.0MB [PGP](https://openresty.org/download/openresty-1.19.9.1.tar.gz.asc) - 6 August 2021)
```
But build with specified compiler (afl-gcc):

```bash
AFL_USE_ASAN=1 ASAN_OPTIONS=detect_leaks=0:handle_segv=0:alloc_dealloc_mismatch=0 AFL_PATH=/usr/local/lib/afl AS=/usr/local/bin/afl-as QA_RPATHS=$[ 0x0001|0x0002 ] rpmbuild -ba ./SPECS/openresty-1.19.9.1.spec
```
`changed strings in openresty-1.19.9.spec:`
```bash
./configure --prefix=%{resty_prefix}/%{dir_name} \
	--with-cc='/usr/local/bin/afl-gcc' \
	--with-cpp='/usr/local/bin/afl-g++'
```
And after:
```bash
[user@localhost ~]$ sudo rpm -i openresty-1.19.9-1.x86_64.rpm
[user@localhost ~]$ whereis openresty
openresty: /usr/bin/openresty /usr/local/openresty
[user@localhost ~]$ file /usr/local/openresty/bin/openresty
/usr/local/openresty/bin/openresty: symbolic link to "`"/usr/local/openresty/nginx/sbin/nginx"`"
[user@localhost ~]$ /usr/local/openresty/bin/openresty -v
nginx version: openresty/1.19.9.1
[user@localhost ~]$ objdump -S /usr/local/openresty/nginx/sbin/nginx | grep afl | head -n 5
  43f679:       48 8d 15 60 6b b4 00    lea    0xb46b60(%rip),%rdx        # f861e0 <__afl_global_area_ptr@@Base>
  43f7a1:       48 8d 15 38 6a b4 00    lea    0xb46a38(%rip),%rdx        # f861e0 <__afl_global_area_ptr@@Base>
```

**Comment by zhuizhuhaomeng** [comment]
this bug was fixed after 1.19.9.1, you can find the fix in 1.21.4.1rc2.
but there still a bug in 1.21.4.1rc2, will release the 1.21.4.1rc3.
